File: converted/source2/source/zombie/asset/Asset.java
package zombie.asset;

import zombie.asset.Asset.ObserverCallback;
import zombie.asset.Asset.PRIVATE;
import zombie.asset.Asset.State;
import zombie.asset.AssetManager.AssetParams;

public abstract class Asset {
   protected final AssetManager m_asset_manager;
   private AssetPath m_path;
   private int m_ref_count;
   final PRIVATE m_priv = new PRIVATE(this);

   protected Asset(AssetPath var1, AssetManager var2) {
      this.m_ref_count = 0;
      this.m_path = var1;
      this.m_asset_manager = var2;
   }

   public abstract AssetType getType();

   public State getState() {
      return this.m_priv.m_current_state;
   }

   public boolean isEmpty() {
      return this.m_priv.m_current_state == State.EMPTY;
   }

   public boolean isReady() {
      return this.m_priv.m_current_state == State.READY;
   }

   public boolean isFailure() {
      return this.m_priv.m_current_state == State.FAILURE;
   }

   public void onCreated(State var1) {
      this.m_priv.onCreated(var1);
   }

   public int getRefCount() {
      return this.m_ref_count;
   }

   public ObserverCallback getObserverCb() {
      if (this.m_priv.m_cb == null) {
         this.m_priv.m_cb = new ObserverCallback();
      }

      return this.m_priv.m_cb;
   }

   public AssetPath getPath() {
      return this.m_path;
   }

   public AssetManager getAssetManager() {
      return this.m_asset_manager;
   }

   protected void onBeforeReady() {
   }

   protected void onBeforeEmpty() {
   }

   public void addDependency(Asset var1) {
      this.m_priv.addDependency(var1);
   }

   public void removeDependency(Asset var1) {
      this.m_priv.removeDependency(var1);
   }

   int addRef() {
      return ++this.m_ref_count;
   }

   int rmRef() {
      return --this.m_ref_count;
   }

   public void setAssetParams(AssetParams var1) {
   }
}

File: ProjectZomboidSwitch/src/zombie/asset/Asset.cpp

#include "zombie/asset/Asset.h"

namespace zombie::asset {

Asset::Asset(AssetPath path, AssetManager asset_manager)
    : m_asset_manager(asset_manager), m_path(path), m_ref_count(0), m_priv(std::make_shared<PRIVATE>(this)) {}

State Asset::getState() const {
    return m_priv->m_current_state;
}

bool Asset::isEmpty() const {
    return m_priv->m_current_state == State::EMPTY;
}

bool Asset::isReady() const {
    return m_priv->m_current_state == State::READY;
}

bool Asset::isFailure() const {
    return m_priv->m_current_state == State::FAILURE;
}

void Asset::onCreated(State state) {
    m_priv->onCreated(state);
}

int Asset::getRefCount() const {
    return m_ref_count;
}

std::shared_ptr<ObserverCallback> Asset::getObserverCb() {
    if (!m_priv->m_cb) {
        m_priv->m_cb = std::make_shared<ObserverCallback>();
    }
    return m_priv->m_cb;
}

AssetPath Asset::getPath() const {
    return m_path;
}

AssetManager Asset::getAssetManager() const {
    return m_asset_manager;
}

void Asset::onBeforeReady() {
    // No-op by default
}

void Asset::onBeforeEmpty() {
    // No-op by default
}

void Asset::addDependency(Asset& asset) {
    m_priv->addDependency(asset);
}

void Asset::removeDependency(Asset& asset) {
    m_priv->removeDependency(asset);
}

int Asset::addRef() {
    return ++m_ref_count;
}

int Asset::rmRef() {
    return --m_ref_count;
}

void Asset::setAssetParams(const AssetParams& params) {
    // No-op by default
}

} // namespace asset

std::shared_ptr<ObserverCallback> Asset::getObserverCb() {
    if (!m_priv->m_cb) {
        m_priv->m_cb = std::make_shared<ObserverCallback>();
    }
    return m_priv->m_cb;
}

AssetPath Asset::getPath() {
    return m_path;
}

AssetManager Asset::getAssetManager() {
    return m_asset_manager;
}

void Asset::onBeforeReady() {
    // No-op by default
}

void Asset::onBeforeEmpty() {
    // No-op by default
}

void Asset::addDependency(Asset& var1) {
    m_priv->addDependency(var1);
}

void Asset::removeDependency(Asset& var1) {
    m_priv->removeDependency(var1);
}

int Asset::addRef() {
    return ++m_ref_count;
}

int Asset::rmRef() {
    return --m_ref_count;
}

void Asset::setAssetParams(const AssetParams& var1) {
    // No-op by default
}

} // namespace asset
} // namespace zombie
